### ISSUE TITLE
Single word translations

### ISSUE THREAD

**Issue by popstick** [modernmt/modernmt#499]
Hello. I've built a model with around 130,000 parallel lines of data with around 50% additional back translated data. 

It's working quite well. But when I try to translate a single word the response is an entire sentence.

Example:

Source: 'hau'
Actual translation: "I"
Modernmt Translation: "i am I myself myself myself myself myself myself myself myself I am taken"

How would I go about tracking down the cause of this or fine tuning it? I'm a bit lost.

**Comment by nicolabertoldi** [comment]
@popstick 

Surely, the output is wrong.
The most likely reason for the babble is the amount of training data, especially if the vocabulary is large. 
Another reason could be a wrong or incomplete training of the system. How may updates were performed?
The lack of quality of the back-translated data can also have had a bad impact. 

Just for curiosity.
Which language are you translating from?
What kind of data you used for training?
What kind of data you back-translated?

**Comment by popstick** [comment]
The pair is Tetun/English
The data is mostly scraped from the web
The back translation is also monolingual data scraped from news sources. 

I think the data is quite clean and well aligned. And I believe the vocabulary is quite small for the language. Less than 30,000 words.

The training completes quite quickly in about 4 hours. Would it be worth forcing it to go longer?

Here is the output from the end of the training log. Does this give any clues?

| epoch 038:    600 / 634 loss=3.315, nll_loss=1.525, ppl=2.88, wps=18896, ups=2, wpb=10143.148, bsz=337.398, num_updates=24059, lr=0.000203874, gnorm=0.615, clip=0.0$
| epoch 038 | loss 3.317 | nll_loss 1.527 | ppl 2.88 | wps 18910 | ups 2 | wpb 10137.472 | bsz 337.208 | num_updates 24092 | lr 0.000203734 | gnorm 0.617 | clip 0.000$
| epoch 038 | valid on 'valid' subset | loss 3.891 | nll_loss 2.071 | ppl 4.20 | num_updates 24092 | best_loss 3.87889
| epoch 039:    100 / 634 loss=3.258, nll_loss=1.458, ppl=2.75, wps=19421, ups=2, wpb=10182.356, bsz=341.861, num_updates=24193, lr=0.000203308, gnorm=0.603, clip=0.0$
| epoch 039:    200 / 634 loss=3.273, nll_loss=1.475, ppl=2.78, wps=19409, ups=2, wpb=10184.313, bsz=340.975, num_updates=24293, lr=0.000202889, gnorm=0.611, clip=0.0$
2019-08-01 03:36:09,857 [INFO] - Stop criterion: current_loss = 4.326912, previous_loss = 4.326962
2019-08-01 03:36:09,857 [INFO] - Training interrupted by termination policy: validation loss has reached its plateau
2019-08-01 03:36:09,858 [INFO] - Training step "train_nn" completed in 3h 45m 0s
2019-08-01 03:36:09,858 [INFO] - Training step "avg_checkpoints" started
2019-08-01 03:36:09,858 [INFO] - Averaging checkpoints: ['/opt/mmt/runtime/tdt2en/tmp/training/_temp_train/nn_model/checkpoint_38_24000.pt', '/opt/mmt/runtime/tdt2en/$
2019-08-01 03:36:13,996 [INFO] - Training step "avg_checkpoints" completed in 4s
2019-08-01 03:36:15,821 [INFO] - Training step "train" completed in 3h 45m 6s

**Comment by nicolabertoldi** [comment]
@popstick 

The training seems end correctly. 
So again I reconfirm that the problem is due to the lack of data.
Are you sure that the dictionary of the source and target sides of the parallel corpus contain only 30K words?

Did you try to translate longer sentences?
Did you try to evaluate the performance on few hundreds of sentences?


I very quickly give a look on the web at Tetun language
and I noticed that there are may accents and apostrophes in the vocabulary

ModernMT never tested the quality of the tokenization of text for Tetun,
For Instance , I found this parallel sentence
```
Tetun: ha’u han ikan
English: I eat fish
```
where the subject contains an apostrophe.  Is this apostrophe mandatory?  In your example you actually did not use it.

The Tetun tokenized sentence becomes 
```
ha ’ u han ikan
```

This is maybe another partial source of error.

**Comment by popstick** [comment]
Longer sentences work quite well, it’s mainly just single words that give weird results. Even translating the other direction.

Perhaps it is just a lack of data as you say.

The accents and apostrophes are used when writing formally but are often dropped in casual writing. Therefor there is a mix of it all in the data. How best to handle that? Would it be better to standardise it all or leave as is and gather more data.

I’ll have to look into how the tokenising works and the effects it’s having on the Tetun language specifically as I can’t quite grasp how it all works.

**Comment by nicolabertoldi** [comment]
@popstick 

I would suggest to remove apostrophes internal to words, and re-train your system.
